storage/block_storage.go

```diff
@@ -104,6 +113,173 @@ func (b *BlockStorage) Initialize(workers []BlockWorker) {
+	if err != nil {
+		return -1, err
+	}
+
+	if !exists {
+		return -1, ErrOldestIndexMissing
+	}
+
+	index, err := strconv.ParseInt(string(rawIndex), 10, 64)
+	if err != nil {
+		return -1, err
+	}
+
+	return index, nil
+}
+
+// GetOldestBlockIndex returns the oldest block index
+// available in BlockStorage.
+func (b *BlockStorage) GetOldestBlockIndex(
+	ctx context.Context,
+) (int64, error) {
+	dbTx := b.db.NewDatabaseTransaction(ctx, false)
+	defer dbTx.Discard(ctx)
+
+	return b.GetOldestBlockIndexTransactional(ctx, dbTx)
+}
+
+// pruneBlock attempts to prune a single block in a database transaction.
+// If a block is pruned, we return its index.
+func (b *BlockStorage) pruneBlock(
+	ctx context.Context,
+	index int64,
+) (int64, error) {
+	// We create a separate transaction for each pruning attempt so that
+	// we don't hit the database tx size maximum. As a result, it is possible
+	// that we prune a collection of blocks, encounter an error, and cannot
+	// rollback the pruning operations.
+	dbTx := b.db.NewDatabaseTransaction(ctx, true)
+	defer dbTx.Discard(ctx)
+
+	oldestIndex, err := b.GetOldestBlockIndexTransactional(ctx, dbTx)
+	if err != nil {
+		return -1, fmt.Errorf("%w: %v", ErrOldestIndexRead, err)
+	}
+
+	if index < oldestIndex {
+		return -1, ErrNothingToPrune
+	}
+
+	head, err := b.GetHeadBlockIdentifierTransactional(ctx, dbTx)
+	if err != nil {
+		return -1, fmt.Errorf("%w: cannot get head block identifier", err)
+	}
+
+	if oldestIndex > head.Index-syncer.PastBlockSize*2 {
+		return -1, ErrPruningDepthInsufficient
+	}
+
+	blockResponse, err := b.getBlockResponse(
+		ctx,
+		&types.PartialBlockIdentifier{Index: &oldestIndex},
+		dbTx,
+	)
+	if err != nil && !errors.Is(err, ErrBlockNotFound) {
+		return -1, err
+	}
+
+	// If there is an omitted block, we will have a non-nil error. When
+	// a block is omitted, we should not attempt to remove it because
+	// it doesn't exist.
+	if err == nil {
+		blockIdentifier := blockResponse.Block.BlockIdentifier
+
+		for _, tx := range blockResponse.OtherTransactions {
+			if err := b.pruneTransaction(ctx, dbTx, blockIdentifier, tx); err != nil {
+				return -1, fmt.Errorf("%w: %v", ErrCannotPruneTransaction, err)
+			}
+		}
+
+		_, blockKey := getBlockHashKey(blockIdentifier.Hash)
+		if err := dbTx.Set(ctx, blockKey, []byte(""), true); err != nil {
+			return -1, err
+		}
+	}
+
+	// Update prune index
+	if err := b.setOldestBlockIndex(ctx, dbTx, true, oldestIndex+1); err != nil {
+		return -1, fmt.Errorf("%w: %v", ErrOldestIndexUpdateFailed, err)
+	}
+
+	// Commit tx
+	if err := dbTx.Commit(ctx); err != nil {
+		return -1, err
+	}
+
+	return oldestIndex, nil
+}
+
+// Prune removes block and transaction data
+// from all blocks with index <= index. Pruning
+// leaves all keys associated with pruned data
+// but overwrites their data to be empty. If pruning
+// is successful, we return the range of pruned
+// blocks.
+//
+// Prune is not invoked automatically because
```

Comment:
Could use some clarification here--what exactly is meant by "invoked automatically"? I see the `Prune()` function in the StatefulSyncer also uses a `for` loop that continuously loops while there is no error, but the comment says that `Prune will repeatedly attempt to prune BlockStorage`

Reply:
"invoked automatically" here means that `BlockStorage` would start to prune on its own without any invocation. A single call to `Prune()` will prune up to the provided index and then return instead of continuing to prune in the background. `statefulsyncer` calls prune repeatedly to continue to prune the chain.